### PR TITLE
Build Controller CA Bundles

### PR DESCRIPTION
**Pull request by adambkaplan**
* Add CA bundle to BuildControllerConfig
* Create config API object for build controller

JIRA-ID: [DEVEXP-182](https://jira.coreos.com/browse/DEVEXP-182)

### CODE REVIEW

File: openshiftcontrolplane/v1/types.go [hunk 1]
```diff
@@ -177,6 +177,8 @@ type BuildControllerConfig struct {
 
 	BuildDefaults  *BuildDefaultsConfig  `json:"buildDefaults"`
 	BuildOverrides *BuildOverridesConfig `json:"buildOverrides"`
+
+	TrustedCABundle string `json:"trustedCABundle"`
```
Comment (adambkaplan) [inline openshiftcontrolplane/v1/types.go]: Ultimate destination for PEM-encoded CA bundle that is mounted into build pods.

**Comment by adambkaplan** [comment]
Client-go change: openshift/client-go#64

File: config/v1/types.go [hunk 2]
```diff
@@ -80,6 +80,32 @@ type ConfigMapReference struct {
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec holds user-settable values for the build controller configuration
+	Spec BuildSpec `json:"spec,omitempty"`
+}
+
+type BuildSpec struct {
+	// AdditionalTrustedCA is a reference to a ConfigMap containing additional CAs that
+	// should be trusted during builds for image pushes and pulls.
+	AdditionalTrustedCA ConfigMapReference `json:"additionalTrustedCA,omitempty"`
```
Comment (bparees) [inline config/v1/types.go]: please add the config for builddefaulter and buildoverrider while we're introducing this, it'll save me some iteration later.

File: openshiftcontrolplane/v1/types.go (hunk 1 above)
Comment (deads2k) [reply openshiftcontrolplane/v1/types.go]: why `additionalTrustedCA` above and `trustedCABundle` here?  Is it no longer the additional one?  Is it the only?  Is there any way to exclude the system roots?

Also, add godoc for this struct while you're here.  For this field, indicate what empty means.

File: config/v1/types.go (hunk 2 above)
Comment (bparees) [reply config/v1/types.go]: thanks, lgtm pending feedback from @deads2k about the right way to reuse the structs (might need to extract it out to a third package that both build and config can depend on)

File: openshiftcontrolplane/v1/types.go (hunk 1 above)
Comment (bparees) [reply openshiftcontrolplane/v1/types.go]: +1 on naming consistency.

I think we can use AdditionalTrustedCA everywhere.  The CAs will be used in addition to the system certs that exist in the builder image, you cannot remove/replace the system certs.

Comment (adambkaplan) [reply openshiftcontrolplane/v1/types.go]: I have worry about the 3.11 -> 4.0 upgrade path when it comes to the image policy config. For 3.11 this is a reference to a file on disk. For 4.0 the string will contain the actual PEM-encoded content.

Comment (bparees) [reply openshiftcontrolplane/v1/types.go]: My expectation is the migration should take the 3.11 on disk content, put it into a configmap key and point Build.Spec.AdditionalTrustedCA to that configmap key.  (ie as if the admin had set it up from scratch on 4.0).

Comment (deads2k) [reply openshiftcontrolplane/v1/types.go]: > For 4.0 the string will contain the actual PEM-encoded content.

Definitely write godoc.  This should reference to a file on disk.

File: config/v1/types.go (hunk 2 above)
Comment (deads2k) [reply config/v1/types.go]: > @deads2k is the dependency on `build/v1` an issue?

That's fun.  How deep do the references go.  @smarterclayton opinions?

File: openshiftcontrolplane/v1/types.go (hunk 1 above)
Comment (bparees) [reply openshiftcontrolplane/v1/types.go]: > This should reference to a file on disk.

why are we not inlining the content into the config.yaml?

File: config/v1/types.go (hunk 2 above)
Comment (bparees) [reply config/v1/types.go]: @deads2k I think we want this to be a pointer since it's optional (and we need to go fix Image.Spec.AdditionalTrustedCA to also be a pointer).  Agree?  Otherwise the omitempty behavior is not what we want.

File: openshiftcontrolplane/v1/types.go (hunk 1 above)
Comment (adambkaplan) [reply openshiftcontrolplane/v1/types.go]: this makes sense at a high level. There a few details I think will warrant discussion in later PRs.

File: config/v1/types.go (hunk 2 above)
Comment (bparees) [reply config/v1/types.go]: @deads2k does not agree, so leave it as is (not a pointer).

File: config/v1/types.go [hunk 3]
```diff
@@ -80,6 +83,96 @@ type ConfigMapReference struct {
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec holds user-settable values for the build controller configuration
+	Spec BuildSpec `json:"spec,omitempty"`
+}
+
+type BuildSpec struct {
+	// AdditionalTrustedCA is a reference to a ConfigMap containing additional CAs that
+	// should be trusted for image pushes and pulls during builds.
+	AdditionalTrustedCA ConfigMapReference `json:"additionalTrustedCA,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: `	// +optional`
so that openapi is generated correctly.  do that throughout on optional fields

File: config/v1/types.go [hunk 4]
```diff
@@ -80,6 +83,96 @@ type ConfigMapReference struct {
...
+	// source strategy.
+	SourceStrategyDefaults *SourceStrategyDefaults `json:"sourceStrategyDefaults,omitempty"`
+
+	// ImageLabels is a list of docker labels that are applied to the resulting image.
+	// User can override a default label by providing a label with the same name in their
+	// Build/BuildConfig.
+	ImageLabels []buildv1.ImageLabel `json:"imageLabels,omitempty"`
+
+	// NodeSelector is a selector which must be true for the build pod to fit on a node
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: Selectors should use the selector format from 3.2.

File: config/v1/types.go (hunk 3 above)
Comment (adambkaplan) [reply config/v1/types.go]: `Build.Spec` should not be optional then, correct?

File: config/v1/types.go [hunk 5]
```diff
@@ -1,8 +1,11 @@
 package v1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+
+	buildv1 "github.com/openshift/api/build/v1"
```
Comment (deads2k) [inline config/v1/types.go]: I wouldn't do this for the simple type you're pulling in.  It's a large dep fanout and no consumer of this API gains a benefit from the shared type.  Customers will use json/yaml, configobserver is required to serialize manually into unstructured.

File: config/v1/types.go (hunk 3 above)
Comment (bparees) [reply config/v1/types.go]: it's optional, technically you don't have to provide any config.

File: config/v1/types.go [hunk 6]
```diff
@@ -80,6 +81,123 @@ type ConfigMapReference struct {
...
+
+	// Env is a set of default environment variables that will be applied to the
+	// build if the specified variables do not exist on the build
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// SourceStrategyDefaults are default values that apply to builds using the
+	// source strategy.
+	// +optional
+	SourceStrategyDefaults *SourceStrategyDefaults `json:"sourceStrategyDefaults,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: avoid pointers unless they fulfill a need as described here: https://github.com/openshift/api/pull/114#issuecomment-433058509

File: config/v1/types.go [hunk 7]
```diff
@@ -80,6 +81,123 @@ type ConfigMapReference struct {
...
+
+	// ImageLabels is a list of docker labels that are applied to the resulting image.
+	// If user provided a label in their Build/BuildConfig with the same name as one in this
+	// list, the user's label will be overwritten.
+	// +optional
+	ImageLabels []ImageLabel `json:"imageLabels,omitempty"`
+
+	// NodeSelector is a selector which must be true for the build pod to fit on a node
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: pointer

File: config/v1/types.go [hunk 8]
```diff
@@ -80,6 +81,123 @@ type ConfigMapReference struct {
...
+
+	// ImageLabels is a list of docker labels that are applied to the resulting image.
+	// User can override a default label by providing a label with the same name in their
+	// Build/BuildConfig.
+	// +optional
+	ImageLabels []ImageLabel `json:"imageLabels,omitempty"`
+
+	// NodeSelector is a selector which must be true for the build pod to fit on a node
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: This stands out to me.  When would a cluster admin want this as a default, but not an override?  

This feels familiar.  @bparees did we talk about this when it landed the first time?

Comment (bparees) [reply config/v1/types.go]: as i recall the distinction had to do with whether we set the key or not, if it's already set in the build itself.

defaulter:  if the user set the key, we don't touch it.
overrider: we stomp the key

Comment (bparees) [reply config/v1/types.go]: (i'm comfortable with making the case that only override is needed and we get rid of the default path, for now)

File: config/v1/types.go [hunk 9]
```diff
@@ -80,6 +81,123 @@ type ConfigMapReference struct {
...
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type SourceStrategyDefaults struct {
+
+	// Incremental indicates if s2i build strategies should perform an incremental
+	// build or not
+	// +optional
+	Incremental *bool `json:"incremental,omitempty"`
```
[the remark this answers is not in the document]
Comment (bparees) [reply config/v1/types.go]: The problem is this is a defaulter.   We need to distinguish between "user did not want to specify a default for this field" and "the user has requested the default value be X"

if we make the field non-optional and default it to 'false' then we have to treat it as if every single cluster everywhere has an admin who said "i want the default to be incremental=false".  

While that's not strictly a problem today (there is no difference between a cluster with a cluster-wide default of incremental=false, and one with no default, because the builds themselves default to incremental=false anyway), it does cause us to lose user intent and may make migration problematic in the future, if for example we wanted to move to a default incremental behavior of true at the build level.

basically it's a second order issue, i think.

nil -> don't touch the user's build (so either you get what the user asked for, or whatever the default build behavior is, which today is false but in the future could be true, theoretically)
false -> set the value to false in the user's build if the user has not set a value in their build
true -> set the value to true in the user's build if the user has not set a value in their build

File: config/v1/types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply config/v1/types.go]: > avoid pointers unless they fulfill a need as described here: [#114 (comment)](https://github.com/openshift/api/pull/114#issuecomment-433058509)

actually, make the selectors strings and this config becomes really obvious for an admin to deal with.

File: config/v1/types.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply config/v1/types.go]: > nil -> don't touch the user's build (so either you get what the user asked for, or whatever the default build behavior is, which today is false but in the future could be true, theoretically)
> false -> set the value to false in the user's build if the user has not set a value in their build
> true -> set the value to true in the user's build if the user has not set a value in their build


That first one is really hard to follow, but I can see the theoretical value I supposed.  Defaulting bools is awful to both think about an explain.

File: config/v1/types.go (hunk 8 above)
Comment (deads2k) [reply config/v1/types.go]: > (i'm comfortable with making the case that only override is needed and we get rid of the default path, for now)

I think we should do this.

**Comment by adambkaplan** [comment]
@deads2k @bparees Removed `BuildDefaults` per your thread.
I also made `ForcePull` optional, as `false` is the implied default.

**Comment by bparees** [comment]
> @deads2k @bparees Removed BuildDefaults per your thread.

i don't think that was the conclusion of the thread.  We were only discussing the NodeSelector field.

I think we need to have a case by case discussion on the other fields as to whether they make more sense to be provided as defaults, or overrides.

File: config/v1/types.go [hunk 10]
```diff
@@ -80,6 +81,120 @@ type ConfigMapReference struct {
...
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type SourceStrategyDefaults struct {
+
+	// Incremental indicates if s2i build strategies should perform an incremental
+	// build or not
+	// +optional
+	Incremental bool `json:"incremental,omitempty"`
```
Comment (deads2k) [inline config/v1/types.go]: @bparees had a good argument for keeping this one a bool pointer if the default is ever going to be true.  If we don't think the default is ever going to be true then it doesn't matter.

File: openshiftcontrolplane/v1/types.go [hunk 11]
```diff
@@ -253,7 +257,7 @@ type BuildDefaultsConfig struct {
 	ImageLabels []buildv1.ImageLabel `json:"imageLabels,omitempty"`
 
 	// nodeSelector is a selector which must be true for the build pod to fit on a node
-	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
```
Comment (deads2k) [inline openshiftcontrolplane/v1/types.go]: You sure you want this one to be a pointer?  It's not touched by a user, so I'll let you have it this way if you want, but it's not obvious to me why you would want to.

**Comment by deads2k** [comment]
two questions.  lgtm otherwise

/approve

**Comment by deads2k** [comment]
oh, and it could use a squash.

File: openshiftcontrolplane/v1/types.go (hunk 11 above)
Comment (adambkaplan) [reply openshiftcontrolplane/v1/types.go]: This may have been an oversight - non-pointer should be OK.

Comment (adambkaplan) [reply openshiftcontrolplane/v1/types.go]: Also missed updating NodeSelector here: https://github.com/openshift/api/pull/113/files#diff-c8352960961d4fca76c619d0d312d4efR293. `LabelSelector` provides a richer set of options that we're exposing to the user.

File: config/v1/types.go [hunk 12]
```diff
@@ -80,6 +81,120 @@ type ConfigMapReference struct {
...
+
+	// ImageLabels is a list of docker labels that are applied to the resulting image.
+	// User can override a default label by providing a label with the same name in their
+	// Build/BuildConfig.
+	// +optional
+	ImageLabels []ImageLabel `json:"imageLabels,omitempty"`
+
+	// Annotations are annotations that will be added to the build pod
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (bparees) [inline config/v1/types.go]: let's drop this. it was previously need to set tolerations, before tolerations became a first class field.  I don't expect anyone is using it anymore.

File: config/v1/types.go [hunk 13]
```diff
@@ -80,6 +81,120 @@ type ConfigMapReference struct {
...
+	// +optional
+	ImageLabels []ImageLabel `json:"imageLabels,omitempty"`
+
+	// NodeSelector is a selector which must be true for the build pod to fit on a node
+	// +optional
+	NodeSelector metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// Annotations are annotations that will be added to the build pod
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (bparees) [inline config/v1/types.go]: let's drop this. it was previously need to set tolerations, before tolerations became a first class field.  I don't expect anyone is using it anymore.

File: config/v1/types.go (hunk 12 above)
Comment (adambkaplan) [reply config/v1/types.go]: should defaults have the `Tolerations` field, then?

File: config/v1/types.go (hunk 10 above)
Comment (adambkaplan) [reply config/v1/types.go]: per @bparees we're going to drop this for now.

File: config/v1/types.go (hunk 12 above)
Comment (bparees) [reply config/v1/types.go]: no, we don't support "default" tolerations in the core master-config so there would be nothing to map it to.

furthermore there's no way for a user to set tolerations on their build directly, so the idea of "override" vs "default" has no meaning.
